### PR TITLE
Features/20250823 custom geoservers integration

### PR DESCRIPTION
**Pull request by ZuitAMB**
# Pull Request
Adds custom Geoserver Integration
## Description
Allows user to add their own geoserver in the frontend settings page.
User can provide, the geoserver base url, possible user and password for authentication, and optionally a name and description for the agent to use.

The agent can use a new custom geoserver tool, to find layers within this geoserver using a search parameter, optional bounding box and a geoserver name or url in case multiple geoserver backends are configured,

In addition to the the Geoserver tool, the tool configurator was simplified and improved.

And more test cases were added to simplify development and stability in the future.

## Related Issue

Fixes #
* WMTS layer representation not working correctly
* Tool Configuration crashing on some optional parameters and types

## Type of Change

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please describe):

## Checklist

- [X] My code follows the code style of this project
- [X] I have tested my changes locally
- [X] I have updated the documentation accordingly
- [X] My changes don't break existing functionality
- [X] I have added tests that prove my fix/feature works

### CODE REVIEW

File: backend/services/tools/geoserver/custom_geoserver.py [hunk 1]
```diff
@@ -0,0 +1,753 @@
...
+        bounding_box = bounding_box.default
+
+    return _get_custom_geoserver_data(
+        actual_state,
+        tool_call_id,
+        search_term,
+        max_results,
+        backend_name,
+    backend_url,
+    bounding_box,
```
[the remark this answers is not in the document]
Comment (ZuitAMB) [reply backend/services/tools/geoserver/custom_geoserver.py]: Fixes by linting

**Review by mucke2701 — APPROVED** [review]
A lot of file changes. In the future we should make sure that old functionality is still given with end-to-end testing and/or keeping multiple instances.

Copilot:
Here’s a review of the pull request based on the provided diff:

---

## Summary

This PR introduces a significant set of features and improvements for GeoServer integration, focusing on:

- **Backend Model Enhancements:** Adds `name` and `description` fields to GeoServer backends for clearer identification and documentation.
- **Wider GeoServer Support:** Adds support for WCS (Web Coverage Service) alongside existing WMS/WFS/WMTS.
- **Robust Parsing & UI Improvements:** Improves frontend parsing of WMTS/WCS URLs, legend rendering, and adds better heuristics.
- **Enhanced Tool Configuration:** Refactors tool configuration to be safer and more flexible.
- **Comprehensive Testing:** Adds extensive backend test coverage for new features and edge cases.

---

## Detailed Review

### Backend (Python)

#### 1. **Models**
- `GeoServerBackend` now includes `name` and `description` fields (optional), which will help identify and document backends in both the settings and UI.
- `DataType` enum now includes `RASTER` to distinguish raster layers.

#### 2. **GeoServer Tools**
- Introduces new backend services and tools for querying multiple GeoServer endpoints, including WMS, WFS, WMTS, and WCS.
- New tool: `get_custom_geoserver_data` is added and registered.
- Tool configuration logic was overhauled in `utility/tool_configurator.py` to allow for prompt overrides and extra configuration without signature mismatches or docstring errors.

#### 3. **Testing**
- Significant new test coverage:
  - Tests for backend filtering, bounding box intersection, and error handling in GeoServer tool logic.
  - Tests for property and CRS sanitization in parsing routines.
  - Tests for tool configuration logic, including prompt overrides, config merging, and edge cases.
  - WMTS-specific test for sanitizing layer templates.

#### 4. **Dependencies**
- Adds `owslib` and `httpx` to the backend dependencies for GeoServer services.
- Improves `pytest` configuration comments for timeout, making it more portable.

---

### Frontend (TypeScript/React)

#### 1. **Settings UI**
- When adding/editing a GeoServer backend, users can now specify a `name` and `description`.
- The settings page renders these fields clearly in the list, improving usability for multiple backends.

#### 2. **Leaflet Map Enhancements**
- **Layer Handling:** The map now supports WCS layers (in addition to WMS, WMTS, WFS).
- **Legend Rendering:** WCS and WMTS layers now have robust legend URL derivation, using heuristics and fallback logic for a variety of GeoServer URL patterns.
- **URL Parsing:** Improved parsing of WMTS and WCS URLs, including handling for workspaces, tile matrix sets, and legend graphics.
- **WFS/GeoJSON:** Improved CRS detection/conversion for WFS layers; attempts to reproject from EPSG:3857 to EPSG:4326 for correct rendering in Leaflet, with validation.
- **Layer Zoom:** Now supports zooming to WFS layers using bounding boxes.

#### 3. **State Management**
- The settings store now supports the new backend fields (`name`, `description`).

---

## Strengths

- **Feature-Rich:** Dramatically increases support for GeoServer services, including raster data via WCS.
- **Robustness:** Handles backend and layer configuration in a more maintainable, extensible manner.
- **Testing:** Introduces comprehensive, isolated tests for new backend behavior and tool configuration, which should catch many edge cases.
- **UI/UX:** Improves the frontend experience when managing backends and viewing map layers.
- **Backward Compatibility:** Frontend and backend changes respect optional fields; existing functionality should not break.

---

## Suggestions

- **Documentation:** Update project documentation to describe the new backend fields and how to use/manage multiple backends.
- **Frontend Validation:** Consider adding form validation to ensure backend URLs are unique or at least not empty.
- **User Feedback:** When adding/removing backends in the frontend, provide success/error feedback to the user.
- **Performance:** If backends or layers become numerous, consider pagination or search in the settings UI.
- **Security:** Ensure that backend credentials are stored/transmitted securely (not visible in logs, etc.).

---

## Conclusion

**This is an excellent, well-structured PR** that significantly upgrades both backend and frontend GeoServer integration and configuration. The changes are wide-reaching but modular, with solid test coverage. It is ready to merge after confirming all CI checks pass and (optionally) adding some documentation/user feedback improvements.

---

**Great work!**
